### PR TITLE
Integrate obs factory

### PR DESCRIPTION
**Pull request by DavidKang**
We made some adjustments to integrate obs_factory engine into OBS. 
We still need to test the correct workflows and fix rubocop complains.

### CODE REVIEW

File: src/api/app/models/obs_factory/distribution.rb [hunk 1]
```diff
@@ -109,7 +109,7 @@ def staging_projects_all
     def source_version
       Rails.cache.fetch("source_version_for_#{name}", expires_in: 10.minutes) do
         begin
-          p = Xmlhash.parse(ActiveXML::backend.direct_http "/source/#{name}/#{SOURCE_VERSION_FILE}?expand=1")
+          p = Xmlhash.parse(ActiveXML.backend.direct_http "/source/#{name}/#{SOURCE_VERSION_FILE}?expand=1")
```
Comment (mdeniz) [inline src/api/app/models/obs_factory/distribution.rb]: I know that this is already a big PR... but this ActiveXML calls must be move to Backen::Api calls

File: src/api/app/helpers/webui/obs_factory/application_helper.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Webui::ObsFactory::ApplicationHelper
     def openqa_links_helper
-      OpenqaJob.openqa_links_url
+      ObsFactory::OpenqaJob.openqa_links_url
```
Comment (mdeniz) [inline src/api/app/helpers/webui/obs_factory/application_helper.rb]: This change could be in a8158a0418f208a90227ebe41bdfa7e2f51024ac instead of having its own commit

File: src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb [hunk 3]
```diff
@@ -15,8 +15,8 @@ def index
       respond_to do |format|
         format.html do
           @staging_projects = ::ObsFactory::StagingProjectPresenter.sort(@distribution.staging_projects_all)
-          @backlog_requests = Request.with_open_reviews_for(by_group: @distribution.staging_manager, target_project: @distribution.name)
-          @requests_state_new = Request.in_state_new(by_group: @distribution.staging_manager, target_project: @distribution.name)
+          @backlog_requests = ::ObsFactory::Request.with_open_reviews_for(by_group: @distribution.staging_manager, target_project: @distribution.name)
```
Comment (mdeniz) [inline src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: This changes could be in a8158a0418f208a90227ebe41bdfa7e2f51024ac instead of in its own commit

File: src/api/app/models/obs_factory/distribution.rb [hunk 4]
```diff
@@ -219,5 +205,20 @@ def openqa_filter(project)
       return strategy.openqa_filter(project)
     end
 
+    private
```
Comment (mdeniz) [inline src/api/app/models/obs_factory/distribution.rb]: Can you squash this commit with d704381424e6d6674ead88d0e9522bd47a9658ed

**Comment by hennevogel** [comment]
@mdeniz go change some diapers or something! ☝

File: src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+          @requests_state_new = ::ObsFactory::Request.in_state_new(by_group: @distribution.staging_manager, target_project: @distribution.name)
+
+          staging_project = Project.find_by_name("#{@distribution.project}:Staging")
+
+          file = ::Backend::Api::Sources::Package.file(staging_project.name, 'dashboard', 'ignored_requests')
+          unless file.to_s.nil?
+            @ignored_requests = YAML.load(file.to_s)
+          end
+          if !@ignored_requests.nil? and @ignored_requests
```
Comment (Ana06) [inline src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: :scream:

**Comment by DavidKang** [comment]
@bgeuken, I prefer to leave the first commit separated because is only moving obs factory Engine into OBS and it will be easier to review the others commits.

File: src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb (hunk 5 above)
Comment (coolo) [reply src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: This code needs to be rewritten completely in the future anyway as this should move into an obs attribute. Polishing this old functionality to perfection is beyong logic.

**Comment by DavidKang** [comment]
@coolo, thanks :smile_cat:

**Comment by coolo** [comment]
I can setup a Factory clone on build-test

**Comment by bgeuken** [comment]
I am building packages from 
 @DavidKang's branch: https://build.opensuse.org/package/show/home:bgeuken:branches:OBS:Server:Unstable/obs-server

File: src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb (hunk 5 above)
Comment (jberry-suse) [reply src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: The scarier part is the need for `unless file.to_s.nil?`...`to_s.nil?`

Comment (coolo) [reply src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: It's actually no longer needed. Your code was using PackageFile.new - which has a rather strange API as it tries to mimik ActiveRecord::Model.

But @DavidKang changed it to use the file method - and that one simply returns a string or throws an exception.

**Comment by coolo** [comment]
I deployed this and setup a minimal https://build-test.opensuse.org/project/staging_projects/openSUSE:Factory

But if there is no ignored_requests file, it will crash - due to the backend refactoring. So please catch this exception - and we're ready for merge IMO

**Comment by coolo** [comment]
```
I, [2018-06-21T11:54:11.465161 #10305]  INFO -- : [57aef12f-29ac-4402-9eaa-da3202c19afc] [10305:177.22] method=GET path=/project/staging_projects/openSUSE:Factory format=html controller=We
bui::ObsFactory::StagingProjectsController action=index status=500 error='ActiveXML::Transport::NotFoundError: <status code="404">
  <summary>package 'dashboard' does not exist</summary>
  <details>404 package 'dashboard' does not exist</details>
</status>

**Comment by DavidKang** [comment]
@coolo, @bgeuken, @eduardoj and @Ana06, could re-review it again. I made some changes in the commit 6e34451

**Comment by bgeuken** [comment]
@DavidKang The commit looks good to me

**Comment by coolo** [comment]
@Ana06?

File: src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+          staging_project = Project.find_by_name("#{@distribution.project}:Staging")
+          dashboard_package = Package.find_by_project_and_name(staging_project.name, 'dashboard')
+
+          if dashboard_package
+            if dashboard_package.file_exists?('ignored_requests')
+              file = ::Backend::Api::Sources::Package.file(staging_project.name, dashboard_package.name, 'ignored_requests')
+              @ignored_requests = YAML.load(file)
+            end
+          else
+            flash[:notice] = "package dashboard doesn't exist for #{staging_project}"
```
Comment (jberry-suse) [inline src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: A dashboard package is not a requirement and as such this will be annoying for projects without one.

Comment (DavidKang) [reply src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: Okey, I will remove it

File: src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb (hunk 5 above)
Comment (bgeuken) [reply src/api/app/controllers/webui/obs_factory/staging_projects_controller.rb]: > jberry-suse  a day ago
> The scarier part is the need for unless file.to_s.nil?...to_s.nil?

I was also suprised... We [renamed](https://github.com/openSUSE/open-build-service/pull/5183/files#diff-9755b746898a16e5abe51d75c3cecf87R59) it now
